# Importing 

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [5]:
path = r'C:\Users\bhava\Documents\18.02.26_Instacart Basket Analysis'

In [6]:
path

'C:\\Users\\bhava\\Documents\\18.02.26_Instacart Basket Analysis'

# Importing dataset folders

In [7]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [8]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Dropping Columns

df.drop(columns = ['variable'])

In [9]:
df_ords.drop(columns = ['eval_set'])

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
...,...,...,...,...,...,...
3421078,2266710,206209,10,5,18,29.0
3421079,1854736,206209,11,4,10,30.0
3421080,626363,206209,12,1,12,18.0
3421081,2977660,206209,13,1,12,7.0


#  If you wanted to remove the "eval_set" column from df_ords and have the result of that removal be the new df_ords

If you had a variable, x, and you wanted to add a value of “2” to that variable, you could write x + 2. However, this only acted like a calculator. It printed the result of the equation, but the value of x stayed the same. If you wanted to actually add “2” to x and have the result of that equation be the new x, you’d need to write x = x + 2.

In [10]:
df_ords = df_ords.drop(columns = ['eval_set'])

#  alternative to overwriting a dataframe is to create a new version with a different name

df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [11]:
df_ords.value_counts()

order_id  user_id  order_number  order_dow  order_hour_of_day  days_since_prior_order
1         112108   4             4          10                 9.0                       1
2280953   767      13            1          18                 23.0                      1
2280955   119072   10            5          13                 29.0                      1
2280956   135372   72            5          18                 3.0                       1
2280957   73662    6             1          13                 8.0                       1
                                                                                        ..
1140724   131093   21            1          16                 6.0                       1
1140725   85700    21            2          11                 6.0                       1
1140726   201418   25            6          10                 3.0                       1
1140727   188970   11            1          15                 27.0                      1
3421

In [12]:
df_ords.head(10)

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


# to find missing variables  and thier count 

df['variable'].value_counts(dropna = False)

In [13]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

# to find variables existing value that are NOT missing and thier count/frenquecy - remove the dropna funtion
This syntax—df['variable']—is the standard way of accessing a column within a dataframe using pandas.

df['variable'].value_counts

In [14]:
df_ords['days_since_prior_order'].value_counts

<bound method IndexOpsMixin.value_counts of 0           NaN
1          15.0
2          21.0
3          29.0
4          28.0
           ... 
3421078    29.0
3421079    30.0
3421080    18.0
3421081     7.0
3421082    30.0
Name: days_since_prior_order, Length: 3421083, dtype: float64>

# df.rename() function, The final argument, inplace = true, means that you want to overwrite the original column name with this new column name rather than creating a new copy.

df.rename(columns = {'old_name' : 'new_name'}, inplace = True)

In [15]:
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [16]:
df_ords.head()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


In [17]:
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [18]:
df_ords.head()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


In [19]:
df_ords.describe()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.710542e+06,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,9.875817e+05,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,8.552715e+05,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.710542e+06,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.565812e+06,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,3.421083e+06,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


# changing data type to string 

In [20]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [21]:
df_ords.describe()

,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.029782e+05,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,5.953372e+04,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,5.139400e+04,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.026890e+05,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,1.543850e+05,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,2.062090e+05,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


In [22]:
df_ords['order_id'].dtype

dtype('O')

# Importing department dataset

In [23]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [24]:
df_dep.head()

,department_id,1,2,3,4,5,6,7,8,9,...,12,13,14,15,16,17,18,19,20,21
0,department,frozen,other,bakery,produce,alcohol,international,beverages,pets,dry goods pasta,...,meat seafood,pantry,breakfast,canned goods,dairy eggs,household,babies,snacks,deli,missing


# Transpose

In [25]:
df_dep.T

,0
department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta


In [26]:
df_dep_t = df_dep.T

In [27]:
df_dep_t

,0
department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta


# Adding an index

In [28]:
df_dep_t.reset_index()

,index,0
0,department_id,department
1,1,frozen
2,2,other
3,3,bakery
4,4,produce
5,5,alcohol
6,6,international
7,7,beverages
8,8,pets
9,9,dry goods pasta


##  1) Create a new header; 2) Remove the first row in the dataframe; and 3) Add the new header

# take the row under the current header (the first row in the dataframe), use that data to create a new header

In [29]:
new_header = df_dep_t.iloc[0]

# “locate an integer” = iloc.

In [30]:
new_header

0    department
Name: department_id, dtype: object

In [31]:
df_dep_t_new = df_dep_t[1:]

In [32]:
df_dep_t_new

,0
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk


# add the new header to the new df

In [33]:
df_dep_t_new.columns = new_header

In [34]:
new_header

0    department
Name: department_id, dtype: object

In [35]:
df_dep_t_new

department_id,department
1,frozen
2,other
3,bakery
4,produce
5,alcohol
6,international
7,beverages
8,pets
9,dry goods pasta
10,bulk


# Data Dictionary

In [36]:
# Data Dictionary using prods and dep - department_id

In [37]:
df_prods.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
1,2,All-Seasons Salt,104,13,9.3
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1,10.5
4,5,Green Chile Anytime Sauce,5,13,4.3


In [38]:
data_dict = df_dep_t_new.to_dict('index')

In [39]:
data_dict 

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [40]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting 

# object['element'] syntax is called indexing

Indexing refers to retrieving some element from an object, be it a dataframe, list, or otherwise. By using the index operator,[], you’re telling Python to look for the element in the brackets within the object to the left of the brackets (so, in this case, looking for a "department_id" value of 19 within the df_prods dataframe).

In [41]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [42]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [43]:
df_prods[df_prods['department_id']==19]

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5
...,...,...,...,...,...
49666,49662,Bacon Cheddar Pretzel Pieces,107,19,3.6
49669,49665,Super Dark Coconut Ash & Banana Chocolate Bar,45,19,6.9
49670,49666,Ginger Snaps Snacking Cookies,61,19,5.2
49675,49671,Milk Chocolate Drops,45,19,3.0


In [44]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [45]:
df_snacks.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5


# another method to achieve same results - loc looks for a column name

In [46]:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [47]:
df_snacks_2.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5


In [48]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [49]:
df_snacks_3.head()

,product_id,product_name,aisle_id,department_id,prices
0,1,Chocolate Sandwich Cookies,61,19,5.8
15,16,Mint Chocolate Flavored Syrup,103,19,5.2
24,25,Salted Caramel Lean Protein & Fiber Bar,3,19,1.9
31,32,Nacho Cheese White Bean Chips,107,19,4.9
40,41,Organic Sourdough Einkorn Crackers Rosemary,78,19,6.5


# Exporting Data

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

df_dep.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))

# Assignment 4.4

# 3. Another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis is 'user_id' - and changinig it to a string

In [50]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [51]:
df_ords['user_id'].dtype

dtype('O')

In [52]:
df_ords.describe()

,order_number,order_day_of_week,order_hour_of_day,days_since_prior_order
count,3.421083e+06,3.421083e+06,3.421083e+06,3.214874e+06
mean,1.715486e+01,2.776219e+00,1.345202e+01,1.111484e+01
std,1.773316e+01,2.046829e+00,4.226088e+00,9.206737e+00
min,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,5.000000e+00,1.000000e+00,1.000000e+01,4.000000e+00
50%,1.100000e+01,3.000000e+00,1.300000e+01,7.000000e+00
75%,2.300000e+01,5.000000e+00,1.600000e+01,1.500000e+01
max,1.000000e+02,6.000000e+00,2.300000e+01,3.000000e+01


## Renaming days_since_prior_order

In [53]:
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [54]:
df_ords.head()

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


## Finding frequency 

In [55]:
df_ords['order_hour_of_day'].value_counts

<bound method IndexOpsMixin.value_counts of 0           8
1           7
2          12
3           7
4          15
           ..
3421078    18
3421079    10
3421080    12
3421081    12
3421082    14
Name: order_hour_of_day, Length: 3421083, dtype: int64>

In [56]:
df_ords["order_hour_of_day"].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

# Answer 4: The busiest hour is:

10:00 (10 AM) with 288,418 orders

The frequency analysis of the order_hour_of_day variable shows that the busiest hour for placing orders is 10:00, with 288,418 orders. This is the highest count among all hours, closely followed by 11:00 and mid-afternoon hours. The results indicate that customer ordering activity peaks during late morning, suggesting that users are most active around this time of day.

General Insights:

• Ordering activity is heavily concentrated between 9 AM and 4 PM
• Very low activity during late night / early morning (0–6)
• Suggests daytime consumer behavior, not nighttime usage

## Data Dictionary value of 4

In [57]:
print(data_dict.get('4'))

{'department': 'produce'}


# Answer 5: The value 4 represents Produce 

## 6. Subset of breakfast items in products 

In [58]:
df_breakfast_1 =  df_prods[df_prods['department_id']==14]

In [59]:
df_breakfast_1.head()

,product_id,product_name,aisle_id,department_id,prices
27,28,Wheat Chex Cereal,121,14,10.1
33,34,NaN,121,14,12.2
67,68,"Pancake Mix, Buttermilk",130,14,13.7
89,90,Smorz Cereal,121,14,3.9
210,211,Gluten Free Organic Cereal Coconut Maple Vanilla,130,14,3.6


In [60]:
df_breakfast_1.tail()

,product_id,product_name,aisle_id,department_id,prices
49330,49326,Cereal Variety Fun Pack,121,14,9.1
49395,49391,Light and Fluffy Buttermilk Pancake Mix,130,14,2.0
49547,49543,Chocolate Cheerios Cereal,121,14,10.8
49637,49633,Shake 'N Pour Buttermilk Pancake Mix,130,14,14.2
49667,49663,Ultra Protein Power Crunch Peanut Butter N' Ho...,57,14,10.2


In [61]:
df_breakfast_1["department_id"].value_counts()

department_id
14    1116
Name: count, dtype: int64

## 7. Subset for dinner parties: Alcohol, Deli, Beverages and Meat/Seafood which are 5, 20, 7 and 12 respectively

In [62]:
df_breakfast_2 = df_prods.loc[df_prods['department_id'].isin([5, 20, 7, 12])]

In [63]:
df_breakfast_2.head(10)

,product_id,product_name,aisle_id,department_id,prices
2,3,Robust Golden Unsweetened Oolong Tea,94,7,4.5
6,7,Pure Coconut Water With Orange,98,7,4.4
9,10,Sparkling Orange Juice & Prickly Pear Beverage,115,7,8.4
10,11,Peach Mango Juice,31,7,2.8
16,17,Rendered Duck Fat,35,12,17.1
19,20,Pomegranate Cranberry & Aloe Vera Enrich Drink,98,7,6.0
22,23,Organic Turkey Burgers,49,12,8.2
34,35,Italian Herb Porcini Mushrooms Chicken Sausage,106,12,15.1
38,39,Daily Tangerine Citrus Flavored Beverage,64,7,12.5
39,40,Beef Hot Links Beef Smoked Sausage With Chile ...,106,12,22.5


In [64]:
df_breakfast_2.tail(10)

,product_id,product_name,aisle_id,department_id,prices
49642,49638,Organic Superfirm Vacuum Packed Tofu,14,20,9.0
49655,49651,Beef Brisket,122,12,20.7
49659,49655,Apple Cider,98,7,10.7
49665,49661,Porto,134,5,8.2
49674,49670,Beef Summer Sausage,106,12,19.2
49676,49672,Cafe Mocha K-Cup Packs,26,7,6.5
49679,49675,Cinnamon Dolce Keurig Brewed K Cups,26,7,14.0
49680,49676,Ultra Red Energy Drink,64,7,14.5
49686,49682,California Limeade,98,7,4.3
49688,49684,"Vodka, Triple Distilled, Twist of Vanilla",124,5,5.3


In [65]:
df_breakfast_2["department_id"].value_counts()

department_id
7     4365
20    1322
5     1056
12     907
Name: count, dtype: int64

# Answer 8: Count of total items

Breakfast items have a count of 1116 in the department dataset
Alcohol items have a count of 4365 in the department dataset
Deli items have a count of 1322 in the department dataset
Beverage items have a count of 1056 in the department dataset
Meat/Seafood items have a count of 907 in the department dataset

## User ID 1 info

In [66]:
df_ords.head(10)

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


In [67]:
df_ords[df_ords["user_id"] == 1]

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order


In [68]:
df_ords[df_ords["user_id"] == 1].shape

(0, 6)

## Since we changed this earlier to a 'str', we can now change it back to an 'int' to evaluate it

In [69]:
df_ords['user_id'] = df_ords['user_id'].astype('int')

In [70]:
df_ords[df_ords["user_id"] == 1]

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


In [71]:
df_ords[df_ords["user_id"] == 1].shape

(11, 6)

In [72]:
df_ords[df_ords["user_id"] == 1].describe()

,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order
count,11.0,11.000000,11.000000,11.000000,10.000000
mean,1.0,6.000000,2.636364,10.090909,19.000000
std,0.0,3.316625,1.286291,3.477198,9.030811
min,1.0,1.000000,1.000000,7.000000,0.000000
25%,1.0,3.500000,1.500000,7.500000,14.250000
50%,1.0,6.000000,3.000000,8.000000,19.500000
75%,1.0,8.500000,4.000000,13.000000,26.250000
max,1.0,11.000000,4.000000,16.000000,30.000000


In [73]:
df_ords[df_ords["user_id"] == 1]["order_hour_of_day"].value_counts()

order_hour_of_day
8     3
7     3
12    1
15    1
9     1
14    1
16    1
Name: count, dtype: int64

# df_ords[df_ords["user_id"] == 1]["orders_day_of_week"].value_counts()

In [77]:
df_ords[df_ords["user_id"] == 1]["days_since_last_order"].value_counts()

days_since_last_order
14.0    2
15.0    1
21.0    1
29.0    1
28.0    1
19.0    1
20.0    1
0.0     1
30.0    1
Name: count, dtype: int64

In [78]:
df_ords[df_ords["user_id"] == 1]

,order_id,user_id,order_number,order_day_of_week,order_hour_of_day,days_since_last_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0
5,3367565,1,6,2,7,19.0
6,550135,1,7,1,9,20.0
7,3108588,1,8,1,14,14.0
8,2295261,1,9,1,16,0.0
9,2550362,1,10,4,8,30.0


## Answer 8 and 9:

# Basic Profile of user_id 1

Filtering the dataset shows:

Total orders: 11

Variables observed: order timing, day of week, and days between orders

This already tells us the user is not highly active compared to typical Instacart customers.

# Ordering Time Behavior

Most common hours: 7 AM and 8 AM (3 orders each)

Remaining orders spread across daytime hours (9, 12, 14, 15, 16)

This user strongly prefers early morning ordering, suggesting routine based purchasing, likely for daily or household needs rather than impulse buying.

# Day of Week Behavior

Orders by weekday:

Day 4: 4 orders (most frequent)

Day 1: 3 orders

Days 2 and 3: 2 orders each

Purchasing is not random. The user shows clustering on specific weekdays, implying habitual shopping patterns.

# Days Between Orders

Most common gap: 14 days

Other gaps vary widely (15 to 30 days)

One instance of 0 days

This user appears to shop roughly every two weeks, but with some irregularity.


The 0 day interval is the most interesting data point and could indicate: Same day repeat order, Order modification or correction

# Data anomaly or duplicate transaction

This is likely what the data engineers considered “strange.”


## Overview:

User places infrequent orders (11 total)

Strong preference for morning hours

Orders cluster on specific weekdays

Typical reorder cycle around two weeks

Presence of an unusual 0 day reorder gap

User 1 displays consistent behavioral patterns in both time of day and shopping frequency, with one potentially anomalous immediate reorder. This could warrant further investigation to determine whether it reflects genuine customer behavior or a data quality issue.

## Exporting data 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [84]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))

## How to avoid the additional index column you get this with your export (by default) in the future:
To prevent this, you should drop the index during the export phase, not the import phase. Whenever you save a wrangled dataset, always add index=False to your code:
df.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index=False)

In [85]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index=False)